models/shopify__customer_email_cohorts.sql

```diff
@@ -46,24 +65,64 @@ with calendar as (
 
 ), windows as (
 
-    {% set partition_string = 'partition by email, source_relation order by date_month rows between unbounded preceding and current row' %}
+    {% set partition_string = 'partition by ' ~ shopify_partition_by_cols('email', 'source_relation') ~ 'order by date_month rows between unbounded preceding and current row' %}
 
     select
         *,
         sum(total_price_in_month) over ({{ partition_string }}) as total_price_lifetime,
         sum(order_count_in_month) over ({{ partition_string }}) as order_count_lifetime,
         sum(line_item_count_in_month) over ({{ partition_string }}) as line_item_count_lifetime,
-        row_number() over (partition by email, source_relation order by date_month asc) as cohort_month_number
+        row_number() over ( 
+            partition by {{ shopify_partition_by_cols('email', 'source_relation') }}
+            order by date_month asc) 
+            as cohort_month_number
     from orders_joined
-        
-), surrogate_key as (
+
+{% if is_incremental() %}
+), backfill_lifetime_sums as (
+    -- for incremental runs we need to fetch the prior lifetimes to properly continue adding to them
+    select
+        source_relation,
+        email,
+        max(total_price_lifetime) as previous_total_price_lifetime,
+        max(order_count_lifetime) as previous_order_count_lifetime,
+        max(line_item_count_lifetime) as previous_line_item_count_lifetime,
+        max(cohort_month_number) as previous_cohort_month_number
+    from {{ this }}
+    where date_month < {{ shopify.shopify_lookback(from_date="max(date_month)", interval=1, datepart='month') }}
```

Comment:
I agree with the one month lookback. For future implementation and reference, why did you choose 1 month here? Was it due to the grain of the model being email/month grain? Also, was there any consideration to looking back further, or how did you arrive at 1 month being sufficient?

Reply:
Making a note here to chat about this one.

Reply:
Summarizing, this is to keep the aggregates in sync with the date range in the first cte.